File: examples/roots_complex.py
#
# jython examples for jas.
# $Id$
#

import sys;

from java.lang import System

from jas import PolyRing, QQ, DD, CR
from jas import terminate, startLog


# polynomial examples: complex roots over Q

#r = Ring( "Rat(x) L" );
#r = Ring( "Q(x) L" );
r = PolyRing( CR(QQ()), "x", PolyRing.lex );
print "Ring: " + str(r);
print;

#automatic: [one,I,x] = r.gens();
print "one = ", one;
print "I   = ", I;
print "x   = ", x;
print;


f1 = x**3 - 2;

f2 = ( x - I ) * ( x + I ) * ( x - 2 * I ) * ( x + (1,2) * I );

f3 = ( x**3 - 2 * I );

#f = f1 * f2 * f3;
#f = f1 * f2;
#f = f1 * f3;
#f = f2 * f3;
f = f2;

print "f = ", f;
print;

startLog();

t = System.currentTimeMillis();
#R = r.complexRoots(f);
R = f.complexRoots();
t = System.currentTimeMillis() - t;
#print "R = ", [ a.elem.ring for a in R ];
print "R = ", [ a.elem.ring.getRoot() for a in R ];
print "complex roots time =", t, "milliseconds";
print

#terminate();
#sys.exit();

#eps = QQ(1,10) ** (DD().elem.DEFAULT_PRECISION-3); # not too big
eps = QQ(1,10) ** 10;
print "eps = ", eps;

t = System.currentTimeMillis();
R = r.complexRoots(f,eps);
t = System.currentTimeMillis() - t;
#print "R = ", [ str(a) for a in R ];
print "R = ", [ a.elem.decimalMagnitude() for a in R ];
print "complex root refinement time =", t, "milliseconds";
print

#startLog();
terminate();

File: examples/roots_complex.rb
#
# jruby examples for jas.
# $Id$
#

require "examples/jas"

# polynomial examples: complex roots over Q

r = PolyRing.new(CR(QQ()),"x",PolyRing.lex);
puts "Ring: " + str(r);
puts;

one,I,x = r.gens();
puts "one = " + one.to_s;
puts "I   = " + I.to_s;
puts "x   = " + x.to_s;
puts;


f1 = x**3 - 2;

f2 = ( x - I ) * ( x + I ) * ( x - 2 * I ) * ( x + 1/2 * I );

f3 = ( x**3 - 2 * I );

#f = f1 * f2 * f3;
#f = f1 * f2;
#f = f1 * f3;
#f = f2 * f3;
f = f2;

puts "f = " + f.to_s;
puts;

startLog();

t = System.currentTimeMillis();
rr = r.complexRoots(f);
t = System.currentTimeMillis() - t;
#puts "rr = " + str(rr);
puts "rr = " + str(rr.map{ |a| str(a.elem.ring.getRoot()) }); 
puts "complex roots time = " + str(t) + " milliseconds";

#eps = QQ(1,10) ** (BigDecimal::DEFAULT_PRECISION-3);
eps = QQ(1,10) ** 10;
puts "eps = " + str(eps);

t = System.currentTimeMillis();
rr = r.complexRoots(f,eps);
t = System.currentTimeMillis() - t;
#puts "rr = ", [ str(r) for r in rr ];
puts "rr = " + str(rr.map{ |a| str(a.elem.decimalMagnitude()) }); 
puts "complex roots refinement time = " + str(t) + " milliseconds";

#startLog();
terminate();
